src/fs.rs

```diff
@@ -97,7 +112,217 @@ fn write_directory_contents(dir: &Directory, fd: &OwnedFd, repo: &Repository) ->
+                    };
+                    LeafContent::ExternalFile(id, buf.st_size as u64)
+                } else {
+                    LeafContent::InlineFile(Vec::from(data))
+                }
+            }
+            FileType::Symlink => {
+                let target = readlinkat(fd, "", [])?;
+                LeafContent::Symlink(OsString::from_vec(target.into_bytes()))
+            }
+            FileType::CharacterDevice => LeafContent::CharacterDevice(buf.st_rdev),
+            FileType::BlockDevice => LeafContent::BlockDevice(buf.st_rdev),
+            FileType::Fifo => LeafContent::Fifo,
+            FileType::Socket => LeafContent::Socket,
+        };
+        Ok(content)
+    }
+
+    fn read_leaf(&mut self, dirfd: &OwnedFd, name: &OsStr, ifmt: FileType) -> Result<Rc<Leaf>> {
+        let oflags = match ifmt {
+            FileType::RegularFile => OFlags::RDONLY,
+            _ => OFlags::PATH,
+        };
+
+        let fd = openat(
+            dirfd,
+            name,
+            oflags | OFlags::NOFOLLOW | OFlags::CLOEXEC,
+            Mode::empty(),
+        )?;
+
+        let (buf, stat) = self.stat(&fd, ifmt)?;
+
+        if let Some(leafref) = self.inodes.get(&buf.st_ino) {
+            Ok(Rc::clone(leafref))
+        } else {
+            let content = self.read_leaf_content(fd, buf)?;
+            let leaf = Rc::new(Leaf { stat, content });
+            if buf.st_nlink > 1 {
+                self.inodes.insert(buf.st_ino, Rc::clone(&leaf));
+            }
+            Ok(leaf)
+        }
+    }
+
+    pub fn read_directory(&mut self, dirfd: impl AsFd, name: &OsStr) -> Result<Directory> {
+        let fd = openat(
+            dirfd,
+            name,
+            OFlags::RDONLY | OFlags::DIRECTORY | OFlags::NOFOLLOW | OFlags::CLOEXEC,
+            Mode::empty(),
+        )?;
+
+        let (_, stat) = self.stat(&fd, FileType::Directory)?;
+        let mut directory = Directory {
+            stat,
+            entries: vec![],
+        };
+
+        for item in Dir::read_from(&fd)? {
+            let entry = item?;
+            let name = OsStr::from_bytes(entry.file_name().to_bytes());
+
+            if name == "." || name == ".." {
+                continue;
+            }
+
+            let inode = self.read_inode(&fd, name, entry.file_type())?;
+            directory.insert(name, inode);
+        }
+
+        Ok(directory)
+    }
+
+    fn read_inode(&mut self, dirfd: &OwnedFd, name: &OsStr, ifmt: FileType) -> Result<Inode> {
+        if ifmt == FileType::Directory {
+            Ok(Inode::Directory(Box::new(
+                self.read_directory(dirfd, name)?,
+            )))
+        } else {
+            Ok(Inode::Leaf(self.read_leaf(dirfd, name, ifmt)?))
+        }
+    }
+}
+
+pub fn read_from_path(path: &Path, repo: Option<&Repository>) -> Result<FileSystem> {
+    let mut reader = FilesystemReader {
+        repo,
+        inodes: HashMap::new(),
+        st_dev: u64::MAX,
```

Comment:
Ditto re `Option`

Reply:
We could do `Option` here indeed, and it would be cleaner.  I did it the way that I did it because I wanted to compare directly because the common case is that they will indeed be equal.  Excessive micro-optimization, to be sure.